listenbrainz/listenstore/timescale_utils.py

```diff
@@ -1,41 +1,228 @@
+              FROM update_ts u
+              JOIN LATERAL (
+                    SELECT min(listened_at) AS min_listened_ts
+                      FROM listen l
+                     WHERE l.listened_at >= u.min_listened_at
+                       -- new minimum will be greater than the last one
+                       AND l.user_id = u.user_id
+                       AND l.created <= u.created
+                   ) AS new_ts
+                ON TRUE
+        )
+            UPDATE listen_user_metadata lm
+               SET min_listened_at = new_listened_at
+              FROM calculate_new_ts mt
+             WHERE lm.user_id = mt.user_id
+    """
+
+    # check for which users the listen of maximum listened_at was deleted, for those users
+    # calculate new maximum listened_at
+    update_listen_max_ts = """
+        WITH update_ts AS (
+            SELECT user_id, max_listened_at, lm.created
+              FROM listen_delete_metadata dl
+              JOIN listen_user_metadata lm
+             USING (user_id)
+          GROUP BY user_id, max_listened_at, lm.created
+            HAVING max_listened_at = ANY(array_agg(dl.listened_at))
+        ), calculate_new_ts AS (
+            SELECT user_id
+                 , new_ts.max_listened_ts AS new_listened_at
+              FROM update_ts u
+              JOIN LATERAL (
+                    SELECT max(listened_at) AS max_listened_ts
+                      FROM listen l
+                     WHERE l.listened_at <= u.max_listened_at
+                       AND l.user_id = u.user_id
+                       -- new maximum will be lesser than the last one
+                       AND l.created <= u.created
+                   ) AS new_ts
+                ON TRUE
+        )
+            UPDATE listen_user_metadata lm
+               SET max_listened_at = new_listened_at
+              FROM calculate_new_ts mt
+             WHERE lm.user_id = mt.user_id
+    """
+    delete_user_metadata = "DELETE FROM listen_delete_metadata WHERE id <= :max_id"
+
+    with timescale.engine.begin() as connection:
+        result = connection.execute(select_max_id)
+        row = result.fetchone()
+
+        if not row:
+            logger.info("No pending deletes")
+            return
+
+        max_id = row["max_id"]
+        logger.info("Found max id in listen_delete_metadata table: %d", max_id)
+
+        logger.info("Deleting Listens and updating affected listens counts")
+        connection.execute(text(delete_listens_and_update_listen_counts), max_id=max_id)
+
+        logger.info("Update minimum listen timestamp affected by deleted listens")
+        connection.execute(text(update_listen_min_ts))
+
+        logger.info("Update maximum listen timestamp affected by deleted listens")
+        connection.execute(text(update_listen_max_ts))
+
+        logger.info("Clean up listen delete metadata table")
+        connection.execute(text(delete_user_metadata), max_id=max_id)
+
+
+def update_user_listen_data():
```

Comment:
docs!

Reply:
This function never gets called except for tests and the cron job, yes?

Reply:
Yes.